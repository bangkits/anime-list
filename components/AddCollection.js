/** @jsxImportSource @emotion/react */
import { css } from '@emotion/react';
import { TextField } from '@material-ui/core';
import { Button, Typography } from '@mui/material';
import AddCircleIcon from '@mui/icons-material/AddCircle';
import { useState } from 'react';
import SnackbarComponent from './Snackbar';
import { useAppDispatch } from '../reducer/hooks';
import { addCollection } from '../reducer/collection.slice';

const wrapper = css({
  display: 'flex',
  justifyContent: 'space-between',
  alignItems: 'flex-end',
  gap: 10,
  marginBottom: 40,
  maxWidth: 600
});

const snackbarState = {
  show: false,
  type: '',
  messasge: ''
};

export default function AddCollection({ collection }) {
  const [input, setInput] = useState('');
  const [snackbar, setSnackbar] = useState(snackbarState);
  const dispatch = useAppDispatch();

  const handleAddCollection = () => {
    if (!input) {
      setSnackbar({
        show: true,
        type: 'error',
        message: 'Please Input Collection Name'
      });
      return;
    };

    if (collection.find(v => v.name.toLowerCase() == input.toLowerCase())) {
      setSnackbar({
        show: true,
        type: 'error',
        message: 'Collection Name Already Exist'
      });
      return;
    };

    dispatch(addCollection({
      id: new Date().valueOf(),
      name: input
    }));

    setSnackbar({
      show: true,
      type: 'success',
      message: 'Collection Added!'
    });
    setInput('');
  };

  return (
    <div css={wrapper}>
      <TextField
        aria-label="collection-input"
        css={{ width: '100%' }}
        label="input collection name"
        value={input}
        onChange={(e) => setInput(e.target.value.replace(/[^a-zA-Z0-9 ]/g, ''))}
        inputProps={{ maxLength: 30 }}
      />
      <Button
        aria-label="collection-submit"
        variant="contained"
        startIcon={<AddCircleIcon />}
        onClick={handleAddCollection}
      >
        <Typography variant='h5'>
          Add
        </Typography>
      </Button>
      <SnackbarComponent
        show={snackbar.show}
        type={snackbar.type}
        message={snackbar.message}
        onDismiss={() => setSnackbar(snackbarState)}
      />
    </div>
  )
};
