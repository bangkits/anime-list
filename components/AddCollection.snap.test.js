import renderer from 'react-test-renderer';
import AddCollection from './AddCollection';

jest.mock('react-redux', () => ({
  useSelector: jest.fn(),
  useDispatch: () => jest.fn()
}));
jest.mock('./Snackbar', () => 'SnackbarComponent');

describe('AddCollection Component Snapshot', () => {
  it('Renders AddCollection component input', () => {
    const props = {
      open: true,
      onDismiss: jest.fn(),
      onSubmit: jest.fn(),
      title: 'Form Title',
      description: 'Form Desc'
    };

    const tree = renderer
      .create(<AddCollection {...props} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
