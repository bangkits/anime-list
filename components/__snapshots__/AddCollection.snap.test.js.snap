// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddCollection Component Snapshot Renders AddCollection component input 1`] = `
<div
  className="css-oqpufy-AddCollection"
>
  <div
    aria-label="collection-input"
    className="MuiFormControl-root MuiTextField-root css-1vohdb9-AddCollection"
  >
    <label
      className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated"
      data-shrink={false}
    >
      input collection name
    </label>
    <div
      className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
      onClick={[Function]}
    >
      <input
        aria-invalid={false}
        autoFocus={false}
        className="MuiInputBase-input MuiInput-input"
        disabled={false}
        maxLength={30}
        onAnimationStart={[Function]}
        onBlur={[Function]}
        onChange={[Function]}
        onFocus={[Function]}
        required={false}
        type="text"
        value=""
      />
    </div>
  </div>
  <button
    aria-label="collection-submit"
    className="MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium MuiButtonBase-root  css-sghohy-MuiButtonBase-root-MuiButton-root"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiButton-startIcon MuiButton-iconSizeMedium css-1d6wzja-MuiButton-startIcon"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddCircleIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm5 11h-4v4h-2v-4H7v-2h4V7h2v4h4v2z"
        />
      </svg>
    </span>
    <h5
      className="MuiTypography-root MuiTypography-h5 css-ag7rrr-MuiTypography-root"
    >
      Add
    </h5>
  </button>
  <SnackbarComponent
    onDismiss={[Function]}
    show={false}
    type=""
  />
</div>
`;
